### PR TITLE
Rework range chapter

### PR DESCRIPTION
**Pull request by wilzbach**
I am really unhappy about the current explanation of a range.
Probably it makes sense splitting it off into two pages.

This example uses UCFS - this is not by accident. Imho we should use UCFS wherever possible and just explain it at the beginning of the Tour. After all, it's a really nice feature :)

### CODE REVIEW

File: basics/ranges.md [hunk 1]
```diff
@@ -21,30 +21,100 @@ for (auto __rangeCopy = range;
...
+```d
+42.repeat.take(3).writeln; // [42, 42, 42]
+```
+
+## Value vs. Reference types
+
+If the range object is a reference type (e.g. `class` or [`std.range.refRange`](https://dlang.org/phobos/std_range.html#refRange)),
+then the range will be consumed and won't available for subsequent iteration
+(that is unless the loop body breaks before the last loop iteration). If the range object is
+a value type, then a copy of the range will be made and depending on the
```
Comment (BBasile) [inline basics/ranges.md]: then the range will be copied

**Comment by wilzbach** [comment]
I finally had time to rework this PR - it should be a huge improvement over the status quo now.

**Comment by wilzbach** [comment]
In lack of responses, I'm going to merge this now as I think it improves the existing page vastly. Please refer to https://tour.dlang.org/tour/en/basics/ranges to submit improvements. Thanks!

**Comment by mywtfmp3** [comment]
The new chapter makes use of `interface`, which is introduced in chapter 16. Is it better to add a forward reference to interface chapter, or rearrange the chapter?

**Comment by wilzbach** [comment]
> The new chapter makes use of interface, which is introduced in chapter 16.

Sorry didn't consider this. Thanks for catching!

> Is it better to add a forward reference to interface chapter, or rearrange the chapter?

Not sure, generally I would say re-arrange, but we want to show the call stuff as early as possible to avoid boring people with the lame boring things they already know, but as they aren't that many sections apart I would say rearrange. What's your personal opinion?

**Pull request merged**